tests/cne_test.cpp

```diff
@@ -41,3 +43,154 @@ TEST_CASE("CNELogisticRegressionTest", "[CNETest]")
+  REQUIRE(abs(coordinates[1]) == Approx(9.66459).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on Schaffer function N.4.
+ */
+TEST_CASE("CNESchafferFunctionN4Test", "[CNETest]")
+{
+  SchafferFunctionN4 f;
+  CNE optimizer(500, 2000, 0.3, 0.3, 0.3, 1e-7);
+
+  arma::mat coordinates = arma::mat("2; 2");
+  optimizer.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(0).margin(0.1));
+  REQUIRE(abs(coordinates[1]) == Approx(1.25313).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on the Ackley Function.
+ */
+TEST_CASE("CNEAckleyFunctionTest", "[CNETest]")
+{
+  AckleyFunction f;
+  CNE optimizer(500, 2000, 0.3, 0.3, 0.3, 1e-7);
+
+  arma::mat coordinates = arma::mat("3; 3");
+  optimizer.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(0).margin(0.1));
+  REQUIRE(coordinates[1] == Approx(0).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on the Beale Function.
+ */
+TEST_CASE("CNEBealeFunctionTest", "[CNETest]")
+{
+  BealeFunction f;
+  CNE optimizer(500, 2000, 0.3, 0.3, 0.3, 1e-7);
+
+  arma::mat coordinates = arma::mat("3; 3");
+  optimizer.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(3).margin(0.1));
+  REQUIRE(coordinates[1] == Approx(0.5).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on the Goldstein-Price Function.
+ */
+TEST_CASE("CNEGoldsteinPriceFunctionTest", "[CNETest]")
+{
+  GoldsteinPriceFunction f;
+  CNE optimizer(500, 2000, 0.3, 0.3, 0.3, 1e-7);
+
+  arma::mat coordinates = arma::mat("1; 0");
+  optimizer.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(0).margin(0.1));
+  REQUIRE(coordinates[1] == Approx(-1).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on the Levi Function.
+ */
+TEST_CASE("CNELevyFunctionN13Test", "[CNETest]")
+{
+  LevyFunctionN13 f;
+  CNE optimizer(500, 2000, 0.3, 0.3, 0.3, 1e-7);
+
+  arma::mat coordinates = arma::mat("3; 3");
+  optimizer.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(1).margin(0.1));
+  REQUIRE(coordinates[1] == Approx(1).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on the Himmelblau Function.
+ */
+TEST_CASE("CNEHimmelblauFunctionTest", "[CNETest]")
+{
+  HimmelblauFunction f;
+  CNE optimizer(500, 2000, 0.3, 0.3, 0.3, 1e-7);
+
+  arma::mat coordinates = arma::mat("2; 1");
+  optimizer.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(3.0).margin(0.1));
+  REQUIRE(coordinates[1] == Approx(2.0).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on the Three-hump Camel Function.
+ */
+TEST_CASE("CNEThreeHumpCamelFunctionTest", "[CNETest]")
+{
+  ThreeHumpCamelFunction f;
+  CNE optimizer(500, 2000, 0.3, 0.3, 0.3, 1e-7);
+
+  arma::mat coordinates = arma::mat("2; 2");
+  optimizer.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(0).margin(0.1));
+  REQUIRE(coordinates[1] == Approx(0).margin(0.1));
+}
+
+/**
+ * Test the CNE optimizer on Schaffer Function N.2.
+ */
+TEST_CASE("CNESchafferFunctionN2Test", "[CNETest]")
```

Comment:
The test fails on my system in 8/10 cases. I uncommented: https://github.com/mlpack/ensmallen/blob/00e26806b4cb5d68ee60cd479fe14daf019bf2de/tests/main.cpp#L25-L27 and used:

```bash
#!/usr/local/bin/zsh

correct=0
while [[ $correct == 0 ]]; do
  ./ensmallen_tests "[CNETest]";
  correct=$?;
done
```

to see how often the test failed.

Reply:
Let me tweak it a little bit.

Reply:
This is a little strange... The reason is that the test doesn't fail that often on my system, and even on travis CI!

Reply:
I can reproduce this also:

```
$ i=0; while(true); do echo $i; ./ensmallen_tests *CNE*; sleep 1; i=$(($i + 1)); done
0
ensmallen version: 1.15.1 (Wrong Side Of The Road)
armadillo version: 9.400.4 (Surrogate Miscreant)
===============================================================================
All tests passed (20 assertions in 10 test cases)

1
ensmallen version: 1.15.1 (Wrong Side Of The Road)
armadillo version: 9.400.4 (Surrogate Miscreant)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ensmallen_tests is a Catch v2.4.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
CNESchafferFunctionN4Test
-------------------------------------------------------------------------------
/home/ryan/src/ensmallen/tests/cne_test.cpp:65
...............................................................................

/home/ryan/src/ensmallen/tests/cne_test.cpp:73: FAILED:
  REQUIRE( coordinates[0] == Approx(0).margin(0.1) )
with expansion:
  1.2534036021 == Approx( 0.0 )

===============================================================================
test cases: 10 |  9 passed | 1 failed
assertions: 19 | 18 passed | 1 failed

2
ensmallen version: 1.15.1 (Wrong Side Of The Road)
armadillo version: 9.400.4 (Surrogate Miscreant)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ensmallen_tests is a Catch v2.4.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
CNESchafferFunctionN2Test
-------------------------------------------------------------------------------
/home/ryan/src/ensmallen/tests/cne_test.cpp:170
...............................................................................

/home/ryan/src/ensmallen/tests/cne_test.cpp:178: FAILED:
  REQUIRE( coordinates[0] == Approx(0).margin(0.1) )
with expansion:
  1.7723130373 == Approx( 0.0 )

===============================================================================
test cases: 10 |  9 passed | 1 failed
assertions: 19 | 18 passed | 1 failed

3
ensmallen version: 1.15.1 (Wrong Side Of The Road)
armadillo version: 9.400.4 (Surrogate Miscreant)
===============================================================================
All tests passed (20 assertions in 10 test cases)

4
ensmallen version: 1.15.1 (Wrong Side Of The Road)
armadillo version: 9.400.4 (Surrogate Miscreant)
===============================================================================
All tests passed (20 assertions in 10 test cases)

5
ensmallen version: 1.15.1 (Wrong Side Of The Road)
armadillo version: 9.400.4 (Surrogate Miscreant)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ensmallen_tests is a Catch v2.4.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
CNESchafferFunctionN4Test
-------------------------------------------------------------------------------
/home/ryan/src/ensmallen/tests/cne_test.cpp:65
...............................................................................

/home/ryan/src/ensmallen/tests/cne_test.cpp:73: FAILED:
  REQUIRE( coordinates[0] == Approx(0).margin(0.1) )
with expansion:
  1.253188829 == Approx( 0.0 )

===============================================================================
test cases: 10 |  9 passed | 1 failed
assertions: 19 | 18 passed | 1 failed

6
ensmallen version: 1.15.1 (Wrong Side Of The Road)
armadillo version: 9.400.4 (Surrogate Miscreant)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ensmallen_tests is a Catch v2.4.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
CNESchafferFunctionN4Test
-------------------------------------------------------------------------------
/home/ryan/src/ensmallen/tests/cne_test.cpp:65
...............................................................................

/home/ryan/src/ensmallen/tests/cne_test.cpp:73: FAILED:
  REQUIRE( coordinates[0] == Approx(0).margin(0.1) )
with expansion:
  -1.2532679093 == Approx( 0.0 )

===============================================================================
test cases: 10 |  9 passed | 1 failed
assertions: 19 | 18 passed | 1 failed
```